lib/jekyll/inclusion.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+module Jekyll
+  class Inclusion
+    attr_reader :name, :path
+
+    def initialize(site, base, name)
+      @site = site
+      @name = name
+      @path = PathManager.join(base, name)
+    end
+
+    def relative_path
+      @relative_path ||= compute_relative_path
+    end
+
+    def render(context)
+      @template ||= site.liquid_renderer.file(relative_path).parse(content)
+      @template.render!(context)
+    rescue Liquid::Error => e
+      e.template_name  = path
+      e.markup_context = "included " if e.markup_context.nil?
+      raise e
+    end
+
+    def content
+      @content ||= File.read(path, **site.file_read_opts)
+    end
+
+    def inspect
+      "#{self.class} #{relative_path.inspect}"
+    end
+    alias_method :to_s, :inspect
+
+    private
+
+    attr_reader :site
+
+    def dir_path
+      if site.theme && path.start_with?(site.theme.includes_path)
```

Comment:
Is there a better way to express this concept? Why is the theme important? Why is the include path important? What purpose does having this code here serve vs. it being somewhere else? Is there a better place to put this?

Reply:
> Why is the theme important?

There are two possible sources to load an include file from: `site.source` and `site.theme.root`.

If a site is using a theme, there's a 50% chance that an include file is loaded from the theme. Include files in such situations would have a long path e.g. `/users/jo/blog/vendor/bundle/ruby/2.4.0/gems/my_theme-0.1.0/_includes/header.html`.

**Conventionally**, the `relative_path` in such situations would be `path` without the preceding
`/users/jo/blog/vendor/bundle/ruby/2.4.0/gems/my_theme-0.1.0/`. At this point, the relative path would be identical to the `relative_path` for an include file from `site.source`.

To eliminate this confusion, I decided to tag include files from the theme with the theme's dirname. Consequently, the above include file will now have a `relative-path` &mdash;
`my_theme-0.1.0/_includes/header.html` &mdash; instantly discernible when calling `#inspect`

> Why is the include path important?

If we know that `path` starts with `"#{site.theme.root}/_includes", then `dir_path` is without a doubt, `<theme-gem-dirname>/_includes`. Instead of using String substitution to remove the unneeded substrings and allocate duplicates for multiple include files from the theme, we just use a cached string object via `PathManager`.

> Is there a better place to put this?

IMHO, placing it here is best so that it is `private`. Elsewhere it would need to be `public` and would carry overhead for future changes.

Reply:
You took my questions a bit too literally, I believe. In fact, the first question is the most important one:

> Is there a better way to express this concept?

I asked this question because these private methods are the only places where we're worried about handling a theme. In cases like these, I always find it's good to step back and wonder "is there something special that requires these special cases here or have I missed another abstraction?"

tl;dr: the special casing we have here for themes is a smell and I'd like to encourage you to find a way to eliminate the smell with another abstraction. That would probably be best done in a separate PR that would then put this PR on hold.